xoto3/dynamodb/write_versioned/prepare.py

```diff
@@ -106,16 +106,21 @@ def all_items_for_next_attempt(
     table_name_onto_hashable_keys: Dict[str, Set[HashableItemKey]] = {
         table_name: set() for table_name in failed_transaction.tables
     }
-    for (table_name, table_data,) in failed_transaction.tables.items():
+    table_name_onto_key_attributes: Dict[str, Sequence[str]] = {
+        table_name: [] for table_name in failed_transaction.tables
+    }
+    for (table_name, table_data) in failed_transaction.tables.items():
         items, effects, key_attributes = table_data
+        table_name_onto_key_attributes[table_name] = key_attributes
         for hashable_item_key in items.keys():
             table_name_onto_hashable_keys[table_name].add(hashable_item_key)
         for hashable_item_key in effects.keys():
             table_name_onto_hashable_keys[table_name].add(hashable_item_key)
 
     return {
         table_name: [
-            hashable_key_to_key(key_attributes, hashable_key) for hashable_key in hashable_keys
```

Comment:
Here, `key_attributes` is being set inside the loop above, and so it holds onto the value for the last table in the transaction. That then gets applied to all of the tables.

Reply:
no kidding. 

this is still probably my least favorite thing about Python - variables should not leak out of their block scope (in this case, the for loop) like this. I wonder if there's a linter rule in pylint that would "prohibit" writing code that does this?

thanks for catching and fixing!